book/chapter-01-unix-sockets/01_unix_sockets_intro.md

```diff
@@ -0,0 +1,292 @@
+```
+
+we are using a special python keywork `with` - this is known as a [context manager](https://docs.python.org/3/library/stdtypes.html#typecontextmanager) in python, which allows us to run code before we enter the `with` block (defined by a `__enter__` method), and run code afterwards (defined by a `__exit__` block). This is usually done to make sure we clean up code or states that are easy to forget to write! Often this is used in contexts when interfacing with files or networking where we want to close ports or release resources back to the operating system.
+
+Let's make a quick example:
+
+```python
+class CustomContext():
+    def __enter__(self):
+        print(">>> Entering context")
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        print("<<< Leaving context")
+
+with CustomContext():
+    print("In context")
+```
+
+Try running the above and see what happens!
+
+---
+
+You might have noticed the `__exit__` method takes multiple other variables than `self` - these are used if you want to handle a certain exception that might happen during your context block.
+
+We can create a slightly more complicated example showing this:
+
+```python
+class CustomContext:
+    def __enter__(self):
+        print(">>> Entering context")
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        print("<<< Leaving context")
+        if isinstance(exc_value, FileNotFoundError):
+            print(f"Exception raised in block: {exc_type}")
+            print(f"Exception message: {exc_value}")
+            return True
+
+
+def start_launch_sequence():
+    raise FileNotFoundError("Launch codes not found.")
+
+
+with CustomContext():
+    print("To the moon 🚀🌕")
+    start_launch_sequence()
+
+print("Done.")
+
+```
+
+Looking at the source files we can [see how the python socket class](https://github.com/python/cpython/blob/main/Lib/socket.py#L215-L242) uses the context manager.
+
+> [RealPython](https://realpython.com/python-with-statement) has a nice tutorial going into further depth with context managers that is worth a look (we've used a class based implementation - but there are other means such as a function based approach!).
+
+#### Using Sockets
+
+The setting `socket.AF_UNIX` denotes that we are using Unix address family, and the `socket.SOCK_STREAM` allows us to create the object in stream mode (rather than using a datagram).
+
+Next we can instruct the kernel to bind the socket to a file with the settings we've defined with `bind`. Unlike other operating systems, a core philopsophy in Unix is that [everything is a file](https://en.wikipedia.org/wiki/Everything_is_a_file) - including sockets! The advantage of this is that the same tools we can use to handle "normal" files can be used to interface with other systems:
+
+```python
+sock.bind("/tmp/docode.sock")
+sock.listen()
+```
+
+If we inspect the file generated after running the server file we can see it is a socket type (`file /tmp/docode.sock`).
```

Comment:
we can also point back to why the `rm /tmp/docode.sock` command works for cleaning up the socket